[Serve] Add metrics to the power-of-two-choices scheduler

shrekris-anyscale opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

This change adds two metrics for the power-of-two-choices scheduler:

* `serve_num_scheduling_tasks`: the number of request scheduling tasks in the router
* `serve_num_scheduling_tasks_in_backoff`: the number of request scheduling tasks undergoing backoff

## Related issue number

<!-- For example: "Closes #1234" -->

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [X] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - This change modifies `test_queued_queries_disconnected` to test the metrics.

GeneDer reviewed (approved):
Can probably add some tests to ensure metrics are tracked in various places, but code logic LGTM

edoakes commented on python/ray/serve/_private/router.py:
```diff
@@ -351,6 +380,30 @@ def target_num_scheduling_tasks(self) -> int:
     def curr_replicas(self) -> Dict[str, ReplicaWrapper]:
         return self._replicas
 
+    def _get_actor_name(self) -> str:
+        """Gets the name of the actor where this scheduler runs.
```
router can run in a driver, in which case we shouldn't log a noisy exception when this fails

shrekris-anyscale replied:
Sounds good, I rewrote the logic to directly return an empty string when running in the driver.

shrekris-anyscale commented on python/ray/serve/_private/router.py:
```diff
@@ -358,6 +388,33 @@ def target_num_scheduling_tasks(self) -> int:
     def curr_replicas(self) -> Dict[str, ReplicaWrapper]:
         return self._replicas
 
+    def _get_actor_name_GCS_UNSAFE(self) -> str:
```
@GeneDer @edoakes What do you think of this naming scheme (`_GCS_UNSAFE` suffix)? I wanted to make it very clear that this method isn't safe to call when the GCS is down.

GeneDer replied:
Personally I don't really like mixed casing in function names, would prefer it all lower cased if you really want to add it. Also feel just have that info in the comment should be sufficient which you already did. But not a hard blocker to me 🙂

edoakes replied:
Yea info in the comment is fine -- there are many calls that can fail when the GCS is down and must be handled at the highest level

shrekris-anyscale replied:
Sounds good, I removed the suffix and left the comments as-is.

shrekris-anyscale commented:
I'll follow up with unit tests next week.

edoakes commented on python/ray/serve/_private/utils.py:
```diff
@@ -734,3 +735,13 @@ def get_all_live_placement_group_names() -> List[str]:
     return live_pg_names
+
+
+def in_ray_driver_process_GCS_UNSAFE() -> bool:
```
I don't think this fails when GCS is down (would be very odd if runtime_context depends on the GCS)

shrekris-anyscale replied:
I tested it by running a Serve app on a 2-node Ray cluster with GCS fault tolerance enabled. All Serve replicas started on the worker node. The Serve replicas ran `ray.get_runtime_context()`.

I killed the head node and tried sending requests to the replicas, but the requests hung. I assume that means `ray.get_runtime_context()` doesn't work when the GCS is down.

edoakes commented on python/ray/serve/_private/router.py:
```diff
@@ -358,6 +388,33 @@ def target_num_scheduling_tasks(self) -> int:
+        """
+
+        if in_ray_driver_process_GCS_UNSAFE():
+            return ""
```
let's return a custom string here for better observability. can use `"DRIVER"` maybe? that way the metrics are still reasonable

edoakes commented on python/ray/serve/_private/router.py:
```diff
@@ -331,6 +332,35 @@ def __init__(
+        self._actor_name: str = self._get_actor_name_GCS_UNSAFE()
+
+        self.num_scheduling_tasks_gauge = metrics.Gauge(
+            "serve_num_power_of_two_choices_scheduling_tasks",
+            description=(
+                "The number of power-of-two-choices scheduling tasks "
+                "run by the router."
```
Sorry, just saw, but I'd suggest the dropping `power_of_two_choices` here from the name (and the other metric) as it is exposing implementation details.

shrekris-anyscale replied:
I dropped `power_of_two_choices`.

shrekris-anyscale replied:
The number of scheduling tasks is also an implementation detail. I added these metrics to the documentation, but we may want to remove it. I'm not sure how informative they are to end users.

shrekris-anyscale commented on python/ray/serve/_private/router.py:
```diff
@@ -358,6 +386,39 @@ def target_num_scheduling_tasks(self) -> int:
+                    # NOTE (shrekris-anyscale): this import is delayed because
+                    # putting at the top of the file causes Ray client tests
+                    # to fail.
+                    from ray.util import state
```
I needed to delay this import because it was causing Ray client tests to fail when it was at the top of the file. The tests that failed were:

* `test_quickstart_class`
* `test_quickstart_counter`
* `test_handle_hanging[serve_with_client0]`

For example, [`test_quickstart_class`](https://github.com/ray-project/ray/blob/b82b8643cc89dd87caf074d8e8bf342d0318fa74/python/ray/serve/tests/test_ray_client.py#L136) failed locally with following traceback:

```
Put failed:


―――――――――――――――――――――――――――― test_quickstart_class ―――――――――――――――――――――――――――――

serve_with_client = None

    def test_quickstart_class(serve_with_client):
>       serve.start()
    
        @serve.deployment

test_ray_client.py:137: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../api.py:107: in start
    client = _private_api.serve_start(detached, http_options, dedicated_cpu, **kwargs)
../_private/api.py:310: in serve_start
    controller, controller_name = _start_controller(
../_private/api.py:182: in _start_controller
    controller = ServeController.options(**controller_actor_options).remote(
../../actor.py:645: in remote
    return actor_cls._remote(args=args, kwargs=kwargs, **updated_options)
../../util/tracing/tracing_helper.py:385: in _invocation_actor_class_remote_span
    return method(self, args, kwargs, *_args, **_kwargs)
../../actor.py:769: in _remote
    return client_mode_convert_actor(self, args, kwargs, **actor_options)
../../_private/client_mode_hook.py:184: in client_mode_convert_actor
    return client_actor._remote(in_args, in_kwargs, **kwargs)
../../util/client/common.py:396: in _remote
    return self.options(**option_args).remote(*args, **kwargs)
../../util/client/common.py:602: in remote
    futures = ray.call_remote(self, *args, **kwargs)
../../util/client/api.py:100: in call_remote
    return self.worker.call_remote(instance, *args, **kwargs)
../../util/client/worker.py:556: in call_remote
    task = instance._prepare_client_task()
../../util/client/common.py:587: in _prepare_client_task
    task = self._remote_stub._prepare_client_task()
../../util/client/common.py:407: in _prepare_client_task
    self._ensure_ref()
../../util/client/common.py:377: in _ensure_ref
    self._ref = ray.worker._put_pickled(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ray.util.client.worker.Worker object at 0x13d96c100>
data = b'\x80\x05\x95\x95\xbf\x00\x00\x00\x00\x00\x00\x8c\x1bray.cloudpickle.cloudpickle\x94\x8c\x14_make_skeleton_class\x94\...0\x00j4\x08\x00\x00j\xef\x07\x00\x00j5\x08\x00\x00j\xf3\x07\x00\x00j6\x08\x00\x00uu\x86\x94\x86R0u}\x94\x86\x94\x86R0.'
client_ref_id = b'\xcc\xa9\xa6\x8aN\x9c\xd2@\x1b\x82U?\xc1\x8c\x89\xfe\xb9'
owner = None

    def _put_pickled(
        self, data, client_ref_id: bytes, owner: Optional[ClientActorHandle] = None
    ):
        req = ray_client_pb2.PutRequest(data=data)
        if client_ref_id is not None:
            req.client_ref_id = client_ref_id
        if owner is not None:
            req.owner_id = owner.actor_ref.id
    
        resp = self.data_client.PutObject(req)
        if not resp.valid:
            try:
>               raise cloudpickle.loads(resp.error)
E               RuntimeError: There is no current event loop in thread 'ray_client_server_0'.

../../util/client/worker.py:510: RuntimeError

 python/ray/serve/tests/test_ray_client.py::test_quickstart_class ⨯50%
```

sihanwang41 commented on python/ray/serve/_private/router.py:
```diff
@@ -799,6 +896,8 @@ def maybe_start_scheduling_tasks(self):
                 self._loop.create_task(self.fulfill_pending_requests())
             )
+        if tasks_to_start > 0:
+            self.num_scheduling_tasks_gauge.set(len(self._scheduling_tasks))
```
why do we set same metric into two places?

shrekris-anyscale replied:
I set the metric whenever a task is added or removed.

sihanwang41 commented on python/ray/serve/_private/router.py:
```diff
@@ -781,6 +877,7 @@ async def fulfill_pending_requests(self):
             logger.exception("Unexpected error in fulfill_pending_requests.")
         finally:
             self._scheduling_tasks.remove(asyncio.current_task(loop=self._loop))
+            self.num_scheduling_tasks_gauge.set(len(self._scheduling_tasks))
```
use `self.curr_num_scheduling_tasks`

shrekris-anyscale replied:
Sounds good, I switched to `self.curr_num_scheduling_tasks`.

shrekris-anyscale commented:
I switched the `actor_name` tag with `actor_id` in both metrics. Getting the `actor_name` requires calling Ray State API, and that was raising obscure errors inside `test_grpc` and `test_client`.

E.g.: on [`test_grpc_proxy_on_draining_nodes`](https://buildkite.com/ray-project/premerge/builds/5202):

```
>       raise RuntimeError(message)
E       RuntimeError: The condition wasn't met before the timeout expired. Last exception: Traceback (most recent call last):
E         File "/rayci/python/ray/_private/test_utils.py", line 553, in wait_for_condition
E           if condition_predictor(**kwargs):
E         File "/rayci/python/ray/serve/tests/utils.py", line 153, in ping_grpc_list_applications
E           response, call = stub.ListApplications.with_call(request=request)
E         File "/opt/miniconda/lib/python3.8/site-packages/grpc/_channel.py", line 1043, in with_call
E           return _end_unary_response_blocking(state, call, True, None)
E         File "/opt/miniconda/lib/python3.8/site-packages/grpc/_channel.py", line 910, in _end_unary_response_blocking
E           raise _InactiveRpcError(state)  # pytype: disable=not-instantiable
E       grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
E       	status = StatusCode.UNAVAILABLE
E       	details = "Socket closed"
E       	debug_error_string = "UNKNOWN:Error received from peer  {created_time:"2023-09-07T20:22:34.980939551+00:00", grpc_status:14, grpc_message:"Socket closed"}"
E       >
```

For the sake of time, I switched the tags, and stopped calling the Ray State API.

shrekris-anyscale commented:
@edoakes this change is ready for review. PTAL.

alexeykudinkin commented on python/ray/serve/_private/router.py:
```diff
@@ -539,107 +599,134 @@ async def choose_two_replicas_with_backoff(
-                    extra={"log_to_stderr": False},
-                )
+        try:
+            entered_backoff = False
```
Are you making any changes to this method? It's really hard to understand right now b/c of the shift.

shrekris-anyscale replied:
I'm just adding the `entered_backoff` flag and [this block](https://github.com/ray-project/ray/pull/38199/files#diff-b81ab953bc243bf2c453db721719903571d53fdd13d9607a292a4658951d66f0R715-R729) to process the gauges.

alexeykudinkin commented on python/ray/serve/_private/router.py:
```diff
@@ -402,6 +403,41 @@ def __init__(
+            {
+                "app": self._deployment_id.app,
+                "deployment": self._deployment_id.name,
+                "actor_id": self._actor_id,
```
We can't add actor-id as a tag, this will blow up our cardinality with every actor recycle

shrekris-anyscale replied:
I see, is there a better way to distinguish between routers on different actors on the same node?

alexeykudinkin replied:
Chatted offline:

 - We already have replica-ids tagged on other metrics
 - This metric is not being added to a dashboard (so impact should be bounded)

alexeykudinkin commented on python/ray/serve/_private/router.py:
```diff
@@ -539,107 +599,134 @@ async def choose_two_replicas_with_backoff(
+                await asyncio.sleep(self.backoff_sequence_s[backoff_index])
+                backoff_index = min(backoff_index + 1, len(self.backoff_sequence_s) - 1)
+        finally:
+            if entered_backoff:
```
I don't think we need all that complexity to track retries: instead we can just emit number of retries and that should be sufficient for us to track (we can't track per-actor metrics anyway)

shrekris-anyscale replied:
What do you mean by `emit number of retries`?

alexeykudinkin replied:
Chatted offline:

 - For the current design we actually want to track a number of scheduling tasks
 - We will revisit in a follow-up decoupling scheduling from replicas assignment (and start tracking number of retries instead of number of tasks)

shrekris-anyscale commented:
[The test failures](https://buildkite.com/ray-project/premerge/builds/5365) are unrelated:

<img width="1160" alt="Screen Shot 2023-09-08 at 8 37 32 AM" src="https://github.com/ray-project/ray/assets/92341594/ffecc8d3-5b44-4010-8485-22c092063322">

<img width="1340" alt="Screen Shot 2023-09-08 at 8 37 56 AM" src="https://github.com/ray-project/ray/assets/92341594/1ab56a9f-efa0-4115-9b41-6a85914ed0b1">

shrekris-anyscale commented:
@edoakes This change is ready to merge.
